File: CPP/YanaPOnlineUtil/Stream/MemoryStreamWriter.cpp
#include "MemoryStreamWriter.h"
#include <memory.h>
#include <string.h>
#include "EndianConverter.h"

namespace YanaPOnlineUtil
{
namespace Stream
{

// コンストラクタ
MemoryStreamWriter::MemoryStreamWriter(unsigned int InBufferSize)
	: pBuffer(nullptr)
	, BufferSize(InBufferSize)
	, CurrentPosition(0)
	, bIsError(false)
{
	pBuffer = new char[BufferSize];
}

// デストラクタ
MemoryStreamWriter::~MemoryStreamWriter()
{
	delete[] pBuffer;
}

// intのシリアライズ
bool MemoryStreamWriter::Serialize(int *pData)
{
	int Data = EndianConverter::Convert(*pData);
	return Write(&Data, sizeof(int));
}

// unsigned intのシリアライズ
bool MemoryStreamWriter::Serialize(unsigned int *pData)
{
	unsigned int Data = EndianConverter::Convert(*pData);
	return Write(&Data, sizeof(unsigned int));
}

// shortのシリアライズ
bool MemoryStreamWriter::Serialize(short *pData)
{
	short Data = EndianConverter::Convert(*pData);
	return Write(&Data, sizeof(short));
}

// unsigned shortのシリアライズ
bool MemoryStreamWriter::Serialize(unsigned short *pData)
{
	unsigned short Data = EndianConverter::Convert(*pData);
	return Write(&Data, sizeof(unsigned short));
}

// charのシリアライズ
bool MemoryStreamWriter::Serialize(char *pData)
{
	return Write(pData, sizeof(char));
}

// unsigned charのシリアライズ
bool MemoryStreamWriter::Serialize(unsigned char *pData)
{
	return Write(pData, sizeof(unsigned char));
}

// floatのシリアライズ
bool MemoryStreamWriter::Serialize(float *pData)
{
	float Data = EndianConverter::Convert(*pData);
	return Write(&Data, sizeof(float));
}

// 文字列のシリアライズ
bool MemoryStreamWriter::Serialize(std::string *pData)
{
	// 文字列長.
	unsigned int Length = pData->length();
	if (!Serialize(&Length)) { return false; }

	// 文字列.
	return Write(pData->c_str(), Length);
}


// 書き込み
bool MemoryStreamWriter::Write(const void *pData, unsigned int Size)
{
	if (CurrentPosition + Size > BufferSize)
	{
		// バッファサイズをオーバーしたらエラー
		bIsError = true;
		return false;
	}

	memcpy((pBuffer + CurrentPosition), pData, Size);
	CurrentPosition += Size;
	return true;
}

}
}

File: CSharp/YanaPOnlineUtil/Stream/MemoryStreamWriter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace YanaPOnlineUtil.Stream
{
	/// <summary>
	/// ストリーム書き込み
	/// </summary>
	public class MemoryStreamWriter : IMemoryStream
	{
		/// <summary>
		/// バッファ
		/// </summary>
		public List<byte> Buffer { get; private set; }

		/// <summary>
		/// コンストラクタ
		/// </summary>
		public MemoryStreamWriter()
		{
			Buffer = new List<byte>();
		}
		
		/// <summary>
		/// シリアライズ
		/// </summary>
		/// <param name="Data">データ</param>
		/// <returns>成功したらtrueを返す</returns>
		public bool Serialize(ref int Data)
		{
			byte[] Bytes = BitConverter.GetBytes(Data);
			if(BitConverter.IsLittleEndian)
			{
				Array.Reverse(Bytes);
			}

			Write(Bytes);
			return true;
		}

		/// <summary>
		/// シリアライズ
		/// </summary>
		/// <param name="Data">データ</param>
		/// <returns>成功したらtrueを返す</returns>
		public bool Serialize(ref uint Data)
		{
			byte[] Bytes = BitConverter.GetBytes(Data);
			if (BitConverter.IsLittleEndian)
			{
				Array.Reverse(Bytes);
			}

			Write(Bytes);
			return true;
		}

		/// <summary>
		/// シリアライズ
		/// </summary>
		/// <param name="Data">データ</param>
		/// <returns>成功したらtrueを返す</returns>
		public bool Serialize(ref short Data)
		{
			byte[] Bytes = BitConverter.GetBytes(Data);
			if (BitConverter.IsLittleEndian)
			{
				Array.Reverse(Bytes);
			}

			Write(Bytes);
			return true;
		}

		/// <summary>
		/// シリアライズ
		/// </summary>
		/// <param name="Data">データ</param>
		/// <returns>成功したらtrueを返す</returns>
		public bool Serialize(ref ushort Data)
		{
			byte[] Bytes = BitConverter.GetBytes(Data);
			if (BitConverter.IsLittleEndian)
			{
				Array.Reverse(Bytes);
			}

			Write(Bytes);
			return true;
		}

		/// <summary>
		/// シリアライズ
		/// </summary>
		/// <param name="Data">データ</param>
		/// <returns>成功したらtrueを返す</returns>
		public bool Serialize(ref char Data)
		{
			byte[] Bytes = new byte[1];
			Bytes[0] = (byte) Data;

			Write(Bytes);
			return true;
		}

		/// <summary>
		/// シリアライズ
		/// </summary>
		/// <param name="Data">データ</param>
		/// <returns>成功したらtrueを返す</returns>
		public bool Serialize(ref byte Data)
		{
			byte[] Bytes = new byte[1];
			Bytes[0] = Data;

			Write(Bytes);
			return true;
		}

		/// <summary>
		/// シリアライズ
		/// </summary>
		/// <param name="Data">データ</param>
		/// <returns>成功したらtrueを返す</returns>
		public bool Serialize(ref string Data)
		{
			int Length = Data.Length;
			if(!Serialize(ref Length)) { return false; }

			byte[] Bytes = new byte[Length];
			for(int i = 0; i < Length; i++)
			{
				Bytes[i] = (byte)Data[i];
			}

			Write(Bytes);
			return true;
		}

		/// <summary>
		/// シリアライズ
		/// </summary>
		/// <param name="Data">データ</param>
		/// <param name="Length">バイト長</param>
		/// <returns>成功したらtrueを返す</returns>
		public bool Serialize(ref char[] Data, int Length)
		{
			throw new NotImplementedException();
		}

		/// <summary>
		/// 書き込み
		/// </summary>
		/// <param name="Bytes">バイト配列</param>
		/// <returns>成功したらtrueを返す</returns>
		private void Write(byte[] Bytes)
		{
			Buffer.AddRange(Bytes);
		}
	}
}
